Expose obj image offset and num images to plugins

nparshook opened the pull request:
This PR makes two changes.
1. Expose the base image offset and number of images for an Object via the plugin API.
2. To support the above I've refactored the image allocation/deallocation into ImageTable.

For testing I opened up a park with all objects available and also wrote a toy plugin to check the images.

IntelOrca commented:
I have a WIP branch where I have refactored the image loading code of objects. In this branch, I created two methods in Object.h: `LoadImages()` and `UnloadImages()`.

Would it be possible for you to change `_baseImageId = GetImageTable().AllocateImages();` to `_baseImageId = LoadImages();`?
And `GetImageTable().FreeImages();` to `UnloadImages`.
These methods will need to be implemented in `Object.h`.
`LoadImages` should return an `ImageIndex`.

Also, you might as well change occurances of image offset to base image id for consistency.

As for exposing the base image index in plugins... I kind of get the rationale. But from what I understand in the discord chat, what you actually need are specific image IDs (footpath preview). In which case it would be better to expose those fields, so that plugins do not rely on the image order of the object as much.

Also, not all objects contain images, so it is somewhat specific to only a certain group of object types. Maybe create another interface that only objects containing images implement in `openrct2.d.ts`?

nparshook commented:
1. If I understand correctly you would like me change my `ImageTable::Allocate/FreeImages` functions into `Object::Load/UnloadImages`?
2. Could you clarify `change occurances of image offset to base image id for consistency`, basically where are you talking about?
3. I agree with exposing just specific fields. However, there is definitely more flexibility with exposing these fields on Object. I think it also makes for a cleaner interface. All objects have 0 or more images, and the preview image is always an Object's first image. Thoughts?
4. Maybe instead of having a separate interface, I make the type an optional, `readonly baseImageId: number|null`?

nparshook commented:
Another question. What is the preferred style for addressing feedback in terms of commits? Do you like PRs to be squashed into a single commit or do you like multiple commits per PR?

IntelOrca commented:
1. yes
2. Change `GetBaseImageOffset` to `GetBaseImageId`, `_image_offset` to `_baseImageId` etc.
3.

> All objects have 0 or more images

Obviously, but some object types can not have 0 images. And the preview image being the first one is not always going to be the case. Some object types have multiple preview images. But this point can be skipped r.e. point 4.

4. Why are you not keen on seperate interfaces? What about just a new interface that implements LoadedObject called `LoadedImageObject`?

nparshook commented:
Mostly because I'm new to typescript and wasn't sure about how interfaces really worked (looks like ducktyping). Hence a little hesitant to add another. I'm going to do some more reading tonight and I'll push a commit tomorrow morning addressing your feedback.

Thanks for the help!

IntelOrca commented:
Music objects can now also contain images, otherwise looks fine.

duncanspumpkin commented on src/openrct2/scripting/bindings/object/ScObject.hpp:
```diff
@@ -125,6 +127,26 @@ namespace OpenRCT2::Scripting
         }
 
+        uint32_t base_image_id_get() const
+        {
```
Don't use snake case. Plugins use their own unique style lowerCamelCase with a _get/set suffix.
